lib/IPCPipeReceive.cpp

```diff
@@ -0,0 +1,60 @@
+/** @file IPCPipeReceive.cpp
+ *  @author Bao Vu (bao.vu@unikie.com)
+ *  @brief
+ */
+
+#include "IPCPipeReceive.h"
+
+IPCPipeReceive::~IPCPipeReceive()
+{
+	if (this->pd > 0)
+		close(this->pd);
+	unlink(this->opts.server_name.c_str());
+}
+
+void IPCPipeReceive::init()
+{
+	std::cout << "Waiting for sender... " << std::endl;
+	this->timer.update_all(); // reset timer
+	do
+	{
+		this->pd = open(this->opts.server_name.c_str(), O_RDONLY | O_NONBLOCK);
+		this->timer.update_end();
+	} while (this->pd <= 0 && this->timer.get_duration() < this->timeout);
+
+	if(this->pd <= 0)
+		throw std::runtime_error(
+			"ERROR: Timeout. Waited for [" + this->opts.server_name + "] more than " + std::to_string(this->timeout) + " seconds.");
+
+	this->open_file();
+	this->timer.update_all(); // reset timer
+
+}
+
+void IPCPipeReceive::transfer()
+{
+	std::vector<char> buffer(this->p_msgsize);
+	std::cout << "Receiving..." << std::endl;
+	while (this->timer.get_duration() < this->timeout)
+	{
+		errno = 0;
+		this->info.read_bytes = read(this->pd, buffer.data(), this->p_msgsize);
+		if (errno == 0 && this->info.read_bytes == 0)
+			break;
+		if (this->info.read_bytes > 0)
+		{
+			this->write_file(buffer, this->info.read_bytes);
+			this->timer.update_all();
+		}
+		this->timer.update_end();
```

Comment:
I am not sure how this loop is suppose to work. What is the scenario where timeout will be actually used? Is it for spinning when EAGAIN was returned? It could be checked if error is actually EAGAIN.

Reply:
I removed sleep(1) to shorten the testing time, will add it later.